onnxruntime/core/framework/partial_executor.cc

```diff
@@ -0,0 +1,497 @@
+  // need unique name for the series. number of nodes should be good enough for a subgraph
+  char series_name[MaxSeriesNameLengthInChars] = "MainGraph";
+  if (graph_viewer->IsSubgraph()) {
+    auto s = graph_viewer->ParentNode()->Name().substr(0, MaxSeriesNameLengthInChars - 1);
+    std::copy(s.cbegin(), s.cend(), series_name);
+  }
+
+  diagnostic::marker_series series(series_name);
+#endif
+
+#ifdef ENABLE_NVTX_PROFILE
+  auto& profile_context = profile::Context::GetInstance();
+  const auto tag = profile_context.GetThreadTagOrDefault(std::this_thread::get_id());
+  profile::NvtxRangeCreator forward_range(
+      "Batch-" + tag + " Forward",
+      profile::Color::White);
+  profile::NvtxRangeCreator backward_range(
+      "Batch-" + tag + " Backward",
+      profile::Color::Black);
+#endif
+
+  for (size_t program_counter = state_.GetProgramCounterStart(); program_counter <= state_.GetProgramCounterEnd(); program_counter += 1) {
+    const auto& node_exec_plan = exec_plan_vec[program_counter];
+    if (terminate_flag_) {
+      LOGS(logger, WARNING) << "Exiting due to terminate flag being set to true.";
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Exiting due to terminate flag being set to true.");
+    }
+
+    auto node_index = node_exec_plan.node_index;
+    const auto& node = *graph_viewer.GetNode(node_exec_plan.node_index);
+
+#ifdef CONCURRENCY_VISUALIZER
+    series.write_flag(node.Name().c_str());
+#endif
+
+#ifdef ENABLE_NVTX_PROFILE
+    if (node.Description() != "Backward pass" && !forward_range.IsBeginCalled()) {
+      // Start timing forward pass when encountering the first forward node.
+      forward_range.Begin();
+    } else if (node.Description() == "Backward pass" && !backward_range.IsBeginCalled() && forward_range.IsBeginCalled()) {
+      // Start timing backward pass when encountering the first backward node.
+      // In the meanwhile, forward range ends.
+      forward_range.End();
+      backward_range.Begin();
+    }
+#endif
+
+    auto p_op_kernel = session_state.GetKernel(node_index);
+    if (p_op_kernel->KernelDef().OpName() == "YieldOp") {
+      // Do not execute YieldOp (it is an no-op anyways).
+      // Decrement the reference count of tensors that are not needed beyond this point.
+      // REVEIW(codemzs): The current model assumes the intermediate tensors that are exported
+      // as graph outputs are owned by ORT, the risk of caller freeing the tensor or manipulating tensor 
+      // memory lingers while the tensor is used downstream after the export.
+      VLOGS(logger, 1) << "Releasing node ML values.";
+      ORT_RETURN_IF_ERROR(ReleaseNodeMLValues(frame, seq_exec_plan, node_exec_plan, logger));
+      continue;
+    }
+    // if a kernel has been added in the session state, it better be NON-null.
+    if (p_op_kernel == nullptr)
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Got nullptr from GetKernel for node: ",
+                             node.Name());
+
```

Comment:
This should be between lines 218 and 219 (although I think it's a superfluous check as session state should ensure there are no nullptr kernels)